Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py

```diff
@@ -1396,6 +1396,35 @@ def _create_absorption_input(self, filename, material=None, geometry=None, envir
         # this effectively deletes the metadata only workspace
         AnalysisDataService.addOrReplace(absName, absorptionWS)
 
+        # Set ChemicalFormula, and either SampleMassDensity or Mass, if SampleMassDensity not set
+        if material is not None:
+            if (not material['ChemicalFormula']) and ("SampleFormula" in absorptionWS.run()):
+                material['ChemicalFormula'] = absorptionWS.run()['SampleFormula'].lastValue().strip()
+            if ("SampleMassDensity" not in material or not material['SampleMassDensity']) and ("SampleDensity" in absorptionWS.run()):
+                if (absorptionWS.run()['SampleDensity'].lastValue() != 1.0) and (absorptionWS.run()['SampleDensity'].lastValue() != 0.0):
+                    material['SampleMassDensity'] = absorptionWS.run()['SampleDensity'].lastValue()
+                else:
+                    material['Mass'] = absorptionWS.run()['SampleMass'].lastValue()
+
+        # Set height for computing density if height not set
+        if geometry is None:
+            geometry = {}
+
+        if geometry is not None:
+            if "Height" not in geometry or not geometry['Height']:
+                # Check units - SetSample expects cm
+                if absorptionWS.run()['BL11A:CS:ITEMS:HeightInContainerUnits'].lastValue() == "mm":
```

Comment:
Another way to do this is to get the units and create a variable "conversion" which is "1" if the units are "mm" and "0.1" if the units are "cm". Then you'd only have one line to set the `geometry['Height']` with the `...lastValue() * conversion`.